File: src/core/melody/grammar.js

export class Grammar {

	constructor(gram_obj) {
		this.gram = gram_obj
		this.variant = 0
	}

	set_variant(v) {
		this.variant = v
	}

	// grammar strings in the gram_obj are white-space-separated strings
	static /*string*/ pick_string(/*[float,string]*/ rule, /*float*/ n) {
		// distribute rules on a line and pick
		let low = 0;
		for (let chance of rule) {
			if (n >= low && n < chance[0] + low)
				return chance[1]
			low += chance[0]
		}
		// in case probabilities don't add up precisely, return last item
		return rule[rule.length-1][1]
	}

	/*string*/ get_starting_symb() {
		let rule = this.gram["starting"][this.variant]
		return Grammar.pick_string(rule, Math.random())
	}

	/*string[]*/ apply_to(/*string[]*/ inp) {
		let out = []
		let ruleset = this.gram["ruleset"][this.variant]
		for (let sym of inp) {
			let rule = ruleset[sym] || [["",1]]
			let s = Grammar.pick_string(rule, Math.random())
			out = out.concat(s.split(" "))
		}
		return out
	}

	generate_sequence(size) {
		let seq = []
		let next = [this.get_starting_symb()]
		while (seq.length < size) {
			next = this.apply_to(next)
			seq = seq.concat(next)
			if (next.length == 0) break // prevent infinite loop
		}
		return seq
	}

}

File: back-end/src/grammar.py
from random import random

class Grammar:

	def __init__(self, gram_obj, variant=0):
		self.gram = gram_obj
		self.variant = variant

	@staticmethod
	def pick_string(rule, n):
		"""pick the corresponding string based on the probability `rule` where the random value is `n`"""
		low = 0
		for chance in rule:
			if n >= low and n < chance[0] + low:
				return chance[1]
			low += chance[0]
		# in case probabilities don't add up precisely, return last item
		return rule[-1][1]

	def pick_starting_symb(self):
		rule = self.gram["starting"][self.variant]
		return Grammar.pick_string(rule, random())

	def apply_to(self, inp):
		"""apply grammar once to `inp`"""
		out = []
		ruleset = self.gram["ruleset"][self.variant]
		for sym in inp:
			rule = ruleset.get(sym, [(1,"")])
			s = Grammar.pick_string(rule, random())
			out += s.split(" ")
		return out

	def generate_sequence(self, size):
		"""
		iterativetely apply grammar until reaching `size`
		or until the grammar stops producing symbols
		"""
		seq = []
		next = [self.pick_starting_symb()]
		while len(seq) < size:
			next = self.apply_to(next)
			seq += next
			if len(next) == 0:
				break # prevent infinite loop if grammar stops producing symbols
		return seq[:size]
